### PR TITLE
Allow extern statics with an extern type

### LINKED ISSUE #55239: Unsized extern statics no longer compile
This is likely due to #55004. Consider the following code:

```rust
#![feature(extern_types)]

extern "C" {
    pub type StartFn;
    pub static start: StartFn;
}
```

This code compiles with `rustc 1.31.0-nightly (e7f5d4805 2018-10-18)`, but fails with `rustc 1.31.0-nightly (78ff609d7 2018-10-19)`. Error message from `rustc`:

```
error[E0277]: the size for values of type `StartFn` cannot be known at compilation time
 --> t.rs:5:23
  |
5 |     pub static start: StartFn;
  |                       ^^^^^^^ doesn't have a size known at compile-time
  |
  = help: the trait `std::marker::Sized` is not implemented for `StartFn`
  = note: to learn more, visit <https://doc.rust-lang.org/book/second-edition/ch19-04-advanced-types.html#dynamically-sized-types-and-the-sized-trait>

error: aborting due to previous error

For more information about this error, try `rustc --explain E0277`.
```

I think this code should still be accepted, and I believe #55004 accidentally broke this. One of the purposes of the `extern_types` feature is to better interact with C code like this:

```c
struct SomeOpaqueType;
extern struct SomeOpaqueType SomeGlobal;
```

I think Rust should accept unsized extern statics that use an extern type (possibly wrapped in a transparent newtype).

cc @oli-obk

### PR DESCRIPTION
**Pull request by mjbshaw**
Fixes #55239

### CODE REVIEW

File: src/librustc_typeck/check/wfcheck.rs [hunk 1]
```diff
@@ -340,23 +340,35 @@ fn check_item_fn<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>, item: &hir::Item) {
     })
 }
 
-fn check_item_type<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>, item_id: ast::NodeId, ty_span: Span) {
+fn check_item_type<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
```
Comment (oli-obk) [inline src/librustc_typeck/check/wfcheck.rs]: please use block-indent (like rustfmt would do it) if you are touching the formatting

File: src/librustc_typeck/check/wfcheck.rs [hunk 2]
```diff
@@ -340,23 +340,35 @@ fn check_item_fn<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>, item: &hir::Item) {
...
-        );
+        let mut allow_unsized = false;
+        if allow_foreign_ty {
+            if let TyKind::Foreign(_) = tcx.struct_tail(item_ty).sty {
+                allow_unsized = true;
+            }
+        }
+
+        if !allow_unsized {
+            fcx.register_wf_obligation(item_ty, ty_span, ObligationCauseCode::MiscObligation);
```
Comment (oli-obk) [inline src/librustc_typeck/check/wfcheck.rs]: we need to keep this obligation in general. Only the `Sized` obligation below should be skipped

File: src/librustc_typeck/check/wfcheck.rs [hunk 3]
```diff
@@ -340,23 +340,35 @@ fn check_item_fn<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>, item: &hir::Item) {
...
-            ),
-        );
+        let mut allow_unsized = false;
+        if allow_foreign_ty {
+            if let TyKind::Foreign(_) = tcx.struct_tail(item_ty).sty {
+                allow_unsized = true;
+            }
+        }
+
+        if !allow_unsized {
```
Comment (oli-obk) [inline src/librustc_typeck/check/wfcheck.rs]: can you remove the negation and invert what the variable means? So `forbid_unsized` is one less thing to think about than what `!allow_unsized` means

**Review by oli-obk — CHANGES_REQUESTED** [review]
lgtm, just some nits

File: src/librustc_typeck/check/wfcheck.rs (hunk 2 above)
Comment (mjbshaw) [reply src/librustc_typeck/check/wfcheck.rs]: Done.

File: src/librustc_typeck/check/wfcheck.rs (hunk 3 above)
Comment (mjbshaw) [reply src/librustc_typeck/check/wfcheck.rs]: Done.

File: src/librustc_typeck/check/wfcheck.rs (hunk 1 above)
Comment (mjbshaw) [reply src/librustc_typeck/check/wfcheck.rs]: Done.

**Review by oli-obk — APPROVED** [review]

**Pull request merged**